### PR TITLE
mds: we should wait messenger when MDSDaemon suicide

### PR DESCRIPTION
**Pull request by flywayway**
Signed-off-by: Wei Feng <feng.wei@h3c.com>

### CODE REVIEW

**Comment by jcsp** [comment]
msgr->wait() is called from ceph_mds.cc  -- I don't think you can wait() in suicide() because you might call suicide() while handling a message (i.e. you are inside the messenger handling code, so would deadlock).  Unless I'm missing something?  What case is this meant to handle?

**Comment by flywayway** [comment]
MonClient will failed to authenticate when mds miss keying, then MDSDaemon will suicide and MDSDaemon::init will return (r < 0), ceph_mds.cc will goto shutdown, msgr->wait() is't called before delete.

**Comment by jcsp** [comment]
OK, so the correct thing in that case is probably to call msgr->wait() from the error handling branch in main()

**Comment by flywayway** [comment]
Can you review this? I add msgr->wait() in main() before msgr be deleted.

File: src/ceph_mds.cc [hunk 1]
```diff
@@ -244,6 +244,7 @@ int main(int argc, const char **argv)
   // detection, etc.).  don't bother if it was a suicide.
   if (mds->is_clean_shutdown()) {
     delete mds;
+    msgr->wait();
```
Comment (jcsp) [inline src/ceph_mds.cc]: For the case you're dealing with, you would want to add the msgr->wait() up at line 216, before "goto shutdown".  Adding it here doesn't work (is_clean_shutdown is something completely different)

**Comment by flywayway** [comment]
Can you review this?

**Pull request merged**